packages/core/src/types/captcha.ts

```diff
@@ -137,7 +141,8 @@ export const CaptchasWithIdSchema = z.array(CaptchaWithIdSchema)
 export const CaptchaSolution = z.object({
     captchaId: z.string(),
     solution: z.number().array(),
-    salt: z.string()
+    salt: z.string(),
+    completed_at: z.any().transform((arg) => arg?.toNumber()),
```

Comment:
Ok, so I kinda overlooked that I've set the value in both places... How would we know if the JS timestamp is valid? Couldn't someone just mess with that value?

Reply:
What would be the benefit if they did?

If they set it in the past then their solution will not be valid for as long so we can discard that idea.

If they set it far in the future then `before_ms: u32::try_from(self.env().block_timestamp() - user.last_correct_captcha)` presumably causes an overflow meaning that the function `dapp_operator_last_correct_captcha` fails.

However, an attacker may be waiting for a specific time in future to have this function pass. Therefore, we can minimise this attack by checking in `provider_approve` whether the `block_timestamp` is more than `n` seconds away from `commitment.completed_at`. Then the question would be what to set `n` to?

Or we just remove the timestamp from JS altogether and implement it only in `ink!`, which actually sounds more secure. This sounds like the cleaner approach.

Reply:
Yeah you are right, not seeing many benefits for changing the value. But since we could be missing something crucial, it might be better to use the block timestamp since, as you said, it sounds cleaner and more secure. So should I just remove the Date.now() part and fix any issues that come with that?

Reply:
Yep, that sounds like the way forward for now. We can always add the JS time in later on if we need to.